bcipy/display/paradigm/matrix/display.py

```diff
@@ -114,10 +114,22 @@ def __init__(self,
         self.stim_registry = self.build_grid()
         self.should_prompt_target = should_prompt_target
 
-        self.logger.info(
+        logger.info(
             f"Symbol positions ({display_container.units} units):\n{self.stim_positions}"
         )
-        self.logger.info(f"Matrix center position: {display_container.center}")
+        logger.info(f"Matrix center position: {display_container.center}")
+
+    def build_symbol_set(self, stimuli: StimuliProperties) -> List[str]:
+        """Build the symbol set for the display."""
+        # TODO pass space and other symbols as parameters
+        if stimuli.layout == 'ALP':
+            return alphabet()
+        elif stimuli.layout == 'QWERTY':
+            return qwerty_order()
+        elif stimuli.layout == 'FREQ':
+            return frequency_order()
+        else:
```

Comment:
Would it make sense for `stimuli.layout` to be a `Callable` that could be used directly?

```python
   def init_symbol_positioner(self, stimuli: StimuliProperties) -> List[str]:
        """Build the symbol set for the display."""
        assert stimuli.layout, "Layout is required"
        return stimuli.layout()
```

It would probably be good to separate out the mapping of the string to the function into another place. An enum would be a good option here since it would also provide a discrete set of values.

```python
def alphabet_order(parameters=None, include_path=True):
  return alphabet(parameters, include_path).index


class MatrixLayout(Enum):
    ALP = ('alp', alphabet_order)
    QWERTY = ('qwerty', qwerty_order)
    FREQ = ('freq', frequency_order)

    def __new__(cls, *args):
      """Autoincrements the value of each item added to the enum."""
      value = len(cls.__members__) + 1
      obj = object.__new__(cls)
      obj._value_ = value
      return obj

    def __init__(self, label, fn):
      self.label = label
      self.symbol_positioner = fn
```

Initialized from parameters.json:
```python
layout = MatrixLayout[parameters.get('matrix_keyboard_layout', 'ALP')]
StimuliProperties(layout=layout.symbol_positioner)
```

Reply:
Implementing these concepts is a little confusing right now: a symbol set and an "order" that redefines the alphabet symbol set. What we want with QWERTY is to define positions and order, but the rows/cols parameters also control this, so in effect, it adds some padding values to the grid and enforces the order.  We may want to separate the concepts of Keyboard, Symbol Set, and Ordering. I'm content with the hacky approach I pushed for now since we have other tickets to complete before the next experiment begins, but happy to discuss more before we release 2.0.

Reply:
I agree that it would be useful to have a higher level interface/API for interacting with these items. I like the idea of passing in a Keyboard, for instance. This would encapsulate the concepts of rows, columns, ordering, spacing. However, I think the underlying low level mechanism will be the same.